nexus/db-queries/src/db/queries/next_item.rs

```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
+///         ORDER BY next_item
+///         LIMIT 1
+///     ),
+///     -- If the table _is_ empty, just insert the minimum value.
+///     <min_item>
+/// )
+/// ```
+///
+/// # Which one to use?
+///
+/// One should probably prefer to use this form of the query, if possible. Both
+/// queries appear to read roughly the same amount of data from disk, but the
+/// series-based implementation uses far more memory and runs more slowly.
+///
+/// One downside of this query is that it always allocates the next _smallest_
+/// item that's available. The `NextItem` query lets the caller choose a base
+/// from which to start the search, which is useful in situations where one
+/// would prefer to randomly distribute items rather than sequentially allocate
+/// them.
+#[derive(Debug, Clone, Copy)]
+pub(super) struct NextItemSelfJoined<
+    Table,
+    Item,
+    ItemColumn,
+    ScopeKey = NoScopeKey,
+    ScopeColumn = NoScopeColumn,
+> {
+    table: Table,
+    _d: PhantomData<(Item, ItemColumn, ScopeColumn)>,
+    scope_key: ScopeKey,
+    item_min: Item,
+    item_max: Item,
+}
+
+impl<Table, Item, ItemColumn, ScopeKey, ScopeColumn>
+    NextItemSelfJoined<Table, Item, ItemColumn, ScopeKey, ScopeColumn>
+where
+    // Table is a database table whose name can be used in a query fragment
+    Table: diesel::Table + HasTable<Table = Table> + QueryFragment<Pg> + Copy,
+
+    // Item can be converted to the SQL type of the ItemColumn
+    Item: ToSql<<ItemColumn as Expression>::SqlType, Pg> + Copy,
+
+    // ItemColum is a column in the target table
+    ItemColumn: Column<Table = Table> + Copy,
+
+    // ScopeKey can be converted to the SQL type of the ScopeColumn
+    ScopeKey: ScopeKeyType + ToSql<<ScopeColumn as Expression>::SqlType, Pg>,
+
+    // ScopeColumn is a column on the target table
+    ScopeColumn: ScopeColumnType + Column<Table = Table>,
+
+    // The Postgres backend supports the SQL types of both columns
+    Pg: HasSqlType<<ScopeColumn as Expression>::SqlType>
+        + HasSqlType<<ItemColumn as Expression>::SqlType>,
+{
+    /// Create a new `NextItemSelfJoined` query, scoped to a particular key.
+    pub(super) fn new_scoped(
+        scope_key: ScopeKey,
+        item_min: Item,
+        item_max: Item,
+    ) -> Self {
+        Self {
+            table: Table::table(),
+            _d: PhantomData,
+            scope_key,
+            item_min,
+            item_max,
+        }
+    }
+}
+
+impl<Table, Item, ItemColumn>
+    NextItemSelfJoined<Table, Item, ItemColumn, NoScopeKey, NoScopeColumn>
+where
+    Table: diesel::Table + HasTable<Table = Table> + QueryFragment<Pg> + Copy,
+    Item: ToSql<<ItemColumn as Expression>::SqlType, Pg> + Copy,
+    ItemColumn: Column<Table = Table> + Copy,
+    Pg: HasSqlType<<ItemColumn as Expression>::SqlType>,
+{
+    /// Create a new `NextItemSelfJoined` query, with a global scope.
+    //
+    // NOTE: This is currently used only in tests.
```

Comment:
Do we ever need this outside the crate? `#[cfg(test)]` should let us omit the `dead_code` warning suppression

Reply:
We don't need it out of this crate, but we do use the unscoped version of the original query today. I'll change the lint though, that's better.

Reply:
Changed to `cfg(test)` in 75311d5a7